processing/src/main/java/io/druid/segment/StringDimensionIndexer.java

```diff
@@ -509,7 +499,7 @@ public boolean matches()
       @Override
       public int getValueCardinality()
       {
-        return maxId;
+        return getCardinality();
```

Comment:
This should still be `maxId`. The selector should act as if it's reading a snapshot of the column that existed at the time the selector was created. So, `getValueCardinality` should be frozen to the snapshot.